docs/05-architectural-decisions.md

```diff
@@ -0,0 +1,567 @@
+
+### 2.5 Implementation Checklist
+
+- [ ] Implement R2 S3-compatible client in CLI
+- [ ] Add credential encryption for config.json storage
+- [ ] Create `emma config` commands for managing authentication
+- [ ] Add Wrangler fallback authentication method
+- [ ] Document setup process in deployment guide
+- [ ] Add error messages for authentication failures
+- [ ] Create troubleshooting guide for common auth issues
+
+## 3. Form Schema Versioning Strategy
+
+### 3.1 Problem Statement
+
+As forms evolve, the schema structure may change. We need a clear versioning strategy to:
+- Track changes to individual forms
+- Ensure backwards compatibility
+- Handle schema validation over time
+- Support form rollback if needed
+
+### 3.2 Requirements
+
+- **Change Tracking**: Record what changed and when
+- **Backwards Compatibility**: Existing submissions remain valid
+- **Deployment Safety**: Prevent breaking changes
+- **Developer Experience**: Simple versioning workflow
+- **Audit Trail**: History of form modifications
+
+### 3.3 Decision: Semantic Versioning with Immutable Deployments
+
+**Version Format**: Follow Semantic Versioning (SemVer)
+
+```
+MAJOR.MINOR.PATCH
+
+Examples:
+- 1.0.0 - Initial deployment
+- 1.1.0 - Added optional email field
+- 1.0.1 - Fixed validation message typo
+- 2.0.0 - Removed required name field (breaking change)
+```
+
+**Version Increment Rules**:
+
+1. **MAJOR** (Breaking Changes)
+   - Removing a required field
+   - Changing field type (e.g., text → number)
+   - Changing field ID
+   - Modifying validation that makes existing data invalid
+
+2. **MINOR** (Additive Changes)
+   - Adding new optional fields
+   - Adding new validation to optional fields
+   - Changing labels or help text
+   - Modifying non-breaking UI settings
+
+3. **PATCH** (Non-Functional Changes)
+   - Fixing typos in labels
+   - Adjusting CSS/themes
+   - Updating success messages
+   - Performance improvements
+
+**Schema Structure**:
+
+```yaml
+formId: contact-form-001
+name: Contact Form
+version: 1.2.0              # Current version
+versionHistory:             # Track all versions
+  - version: 1.0.0
+    deployedAt: 2025-10-01T10:00:00Z
+    changes: Initial release
+  - version: 1.1.0
+    deployedAt: 2025-10-08T15:30:00Z
+    changes: Added optional company field
+  - version: 1.2.0
+    deployedAt: 2025-10-16T12:00:00Z
+    changes: Added phone number field with validation
+
+fields:
+  - id: name
+    type: text
+    label: Your Name
+    required: true
+    addedIn: 1.0.0          # Track when field was added
+```
+
+**CLI Workflow**:
+
+```bash
+# Editing a form prompts for version bump
+emma edit contact-form-001
+# CLI detects changes and suggests version bump:
+# "You removed a required field. Suggested version: 2.0.0 (major)"
+
+# Manual version specification
+emma deploy cloudflare contact-form-001 --version 1.2.0
+
+# View version history
+emma versions contact-form-001
+```
+
+### 3.4 Version Storage in Database
```

Comment:
I want to avoid as much as possible to tight form manage to the database, meaning if we can have a solution to only rely on the bucket I would be happy. Maybe a meta description file or asset is better, it should be barely ever read or written so database is not much a benefit here.